ethers-core/src/types/transaction/response.rs

```diff
@@ -508,10 +498,59 @@ impl PartialOrd<Self> for TransactionReceipt {
     }
 }
 
+impl TransactionReceipt {
+    /// Encode a transaction receipt into bytes base on RLP.
+    ///
+    /// According to [`EIP-2718`]:
+    /// - `Receipt` is either `TransactionType || ReceiptPayload` or `LegacyReceipt`.
+    /// - `LegacyReceipt` is kept to be RLP encoded bytes; it is `rlp([status, cumulativeGasUsed,
+    ///   logsBloom, logs])`.
+    /// - `ReceiptPayload` is an opaque byte array whose interpretation is dependent on the
+    ///   `TransactionType` and defined in future EIPs.
+    ///   - As [`EIP-2930`] defined: if `TransactionType` is `1`, `ReceiptPayload` is `rlp([status,
+    ///     cumulativeGasUsed, logsBloom, logs])`.
+    ///   - As [`EIP-1559`] defined: if `TransactionType` is `2`, `ReceiptPayload` is `rlp([status,
+    ///     cumulative_transaction_gas_used, logs_bloom, logs])`.
+    ///
+    /// [`EIP-2718`]: https://eips.ethereum.org/EIPS/eip-2718#receipts
+    /// [`EIP-2930`]: https://eips.ethereum.org/EIPS/eip-2930#parameters
+    /// [`EIP-1559`]: https://eips.ethereum.org/EIPS/eip-1559#specification
+    pub fn rlp(&self) -> Bytes {
+        let mut rlp = RlpStream::new();
+        rlp.begin_list(4);
+        rlp_opt(&mut rlp, &self.status);
+        rlp.append(&self.cumulative_gas_used);
+        rlp.append(&self.logs_bloom);
+        rlp.append_list(&self.logs);
+
+        let rlp_bytes: Bytes = rlp.out().freeze().into();
+        let mut encoded = vec![];
+        match self.transaction_type {
+            Some(x) if x == U64::from(0x1) => {
+                encoded.extend_from_slice(&[0x1]);
+                encoded.extend_from_slice(rlp_bytes.as_ref());
+                encoded.into()
+            }
+            Some(x) if x == U64::from(0x2) => {
+                encoded.extend_from_slice(&[0x2]);
+                encoded.extend_from_slice(rlp_bytes.as_ref());
+                encoded.into()
+            }
+            #[cfg(feature = "optimism")]
+            Some(x) if x == U64::from(0x7E) => {
+                encoded.extend_from_slice(&[0x7E]);
+                encoded.extend_from_slice(rlp_bytes.as_ref());
+                encoded.into()
+            }
+            _ => rlp_bytes,
+        }
```

Comment:
Might also be worth encoding the type always if `<= 0x7f`?

```suggestion
        match self.transaction_type.map(|x| x.as_u64()) {
            Some(0x1 | 0x2) => return [&[ty as u8][..], &rlp_bytes[..]].concat().into(),
            #[cfg(feature = "optimism")]
            Some(0x7E) => return [&[ty as u8][..], &rlp_bytes[..]].concat().into(),
            _ => {}
        }
        rlp_bytes.into()
```